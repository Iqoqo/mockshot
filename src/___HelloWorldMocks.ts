export class HelloWorldMocks {
    static bar;

    static bar(mock: "no-valid-params" | "success"): any {
        switch (mock) {
            case "no-valid-params":
                return {
                  "error": "yes"
                }
            case "success":
                return {
                  "api": {
                    "data": "123"
                  }
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static shape;

    static shape(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "data": {
                    "_id": "1234"
                  },
                  "id": "1234sddd56",
                  "some": "value"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static foo;

    static foo(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "foo": "barr"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }
}
