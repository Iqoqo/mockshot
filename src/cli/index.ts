import { ApiGenerator } from "../generators/ApiGenerator";
import { GeneratorRunner } from "../GeneratorRunner";
import { SnapshotFetcher } from "./util";

export async function run(args: string[]) {
  const outputDir = args[0];

  const snapshots = SnapshotFetcher.getSnapshots();

  const generators = [new ApiGenerator()];

  const generatorRunner = new GeneratorRunner(outputDir, generators, snapshots);
  await generatorRunner.run();
}
