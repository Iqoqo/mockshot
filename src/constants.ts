export const MockshotTag = "[mockshot]";
export const Success = "success";
