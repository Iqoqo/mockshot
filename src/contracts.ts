export interface ISnapshot {
  key: string;
  packageName?: string;
  filePath: string;
  data: any;
}

export interface IApiSnapshot extends ISnapshot {
  data: IApiSnapData;
}

export interface IClassSnapshot extends ISnapshot {
  data: IClassSnapData;
}

export interface IApiSnapData extends IApiSnapDataBase {
  mockName: string;
}

export interface IClassSnapData {
  className: string;
  methodName: string;
  mockName: string;
  mock: any;
}

export interface IApiSnapDataBase {
  url: string;
  httpMethod: "post" | "get" | "put" | "delete" | "patch";
  mock: {
    statusCode: number;
    body: any;
    error?: string;
  };
}

export type MatcherReturn = {
  pass: boolean;
  message(): string | (() => string);
};
