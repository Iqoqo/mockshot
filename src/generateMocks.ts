import {glob} from "glob";
import * as path from "path";
import Project from "ts-simple-ast";
import fs from "fs";
import { SnapshotState } from "jest-snapshot";
import pretty from 'json-pretty'

export async function generateMocks() {
  const project = new Project({
    compilerOptions: { outDir: "dist/mocks", declaration: true }
  });

  glob("**/*.snap", async function(er, files) {
    // files is an array of filenames.
    // If the `nonull` option is set, and nothing
    // was found, then files is ["**/*.js"]
    // er is an error object or null.
    const mockDef = {} as any;

    files.forEach(file => {
      const absolutePath = path.resolve(file);
      const state = new SnapshotState(absolutePath, {
        snapshotPath: absolutePath
      });
      const data = state._snapshotData;
      const keys = Object.keys(data);

      keys.forEach(async key => {
        if (key.indexOf("[mockshot]") < 0) {
          return;
        }
        const snapshot = JSON.parse(state._snapshotData[key]);
        let classDef = mockDef[snapshot.className];
        if (!classDef) {
          classDef = {};
          mockDef[snapshot.className] = classDef;
        }
        let methodDef = classDef[snapshot.methodName];
        if (!methodDef) {
          methodDef = {};
          classDef[snapshot.methodName] = methodDef;
        }

        let mockName = methodDef[snapshot.mockName];
        if (!mockName) {
          methodDef[snapshot.mockName] = snapshot.mock;
        } else {
          throw Error(
            "Duplicate mock name for method " +
              snapshot.methodName +
              ": " +
              snapshot.mockName
          );
        }
      });
    });

    const classNames = Object.keys(mockDef);
    classNames.forEach(async className => {
      const mockClassName = className+'Mocks';

      console.log('Creating class', mockClassName);

      const mockFileName = `mocks/${mockClassName}.ts`;

      let myClassFile;
      try {
        await fs.unlinkSync(mockFileName);
        console.log("File exists", mockFileName, "removing...");
      } catch (ex) {
      }

      myClassFile = project.createSourceFile(mockFileName);

      const classDeclaration = myClassFile.addClass({
        name: mockClassName
      });

      classDeclaration.setIsExported(true);

      const methods = mockDef[className];

      const methodNames = Object.keys(methods);

      methodNames.forEach(methodName => {
        console.log('Creating method mocks', methodName);

        const mocks = methods[methodName];
        const mockNames = Object.keys(mocks);
        let mockTypes = mockNames.map(value => {
          return `"${value}"`;
        });
        const types = mockTypes.join(" | ");

        const method = classDeclaration.addMethod({
          isStatic: true,
          parameters: [{ name: "mock", type: types }],
          name: methodName,
          returnType: "any"
        });

        method.setBodyText(writer =>
          writer.write("switch (mock)").block(() => {
            mockNames.forEach(mockName => {
              writer.write(`case "${mockName}":`).indentBlock(()=>{
                writer.write(`return ${pretty(mocks[mockName])}`);
              })
            });
            writer.write(`default:`).indentBlock(()=>{
              writer.write(`throw Error("Unknown mock: "+mock);`);
            })
          })
        );
      });
    });

    await project.save();

    console.log('Saved Typescript source...emitting vanilla js');
    await project.emit();
    console.log('Done!');
  });
}
