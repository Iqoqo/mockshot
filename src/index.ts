export * from "./matchers";
export * from "./generators";
export { run } from "./run";
