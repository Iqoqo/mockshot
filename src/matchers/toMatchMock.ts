import { cloneDeep, get, set } from "lodash";
import pretty from "json-pretty";

declare global {
  namespace jest {
    // tslint:disable-next-line:interface-name
    interface Matchers<R> {
      toMatchMock(
        className: string,
        methodName: string,
        mockName: string,
        ignoredKeyPaths?: string[]
      ): R;
    }
  }
}

expect.addSnapshotSerializer({
  test: val => val.mock,
  print: val => pretty(val)
});

function toMatchMock(
  mock,
  className: string,
  methodName: string,
  mockName: string,
  ignoredKeyPaths?: string[]
) {
  const snapshotTag = `[mockshot] [[${className} ${methodName} ${mockName}]]`;
  const snapshotName = `${this.currentTestName}: ${snapshotTag} 1`;
  const currentSnapshot = this.snapshotState._snapshotData[snapshotName];

  if (ignoredKeyPaths && currentSnapshot) {
    const parsedSnapshot = JSON.parse(currentSnapshot);

    ignoredKeyPaths.forEach(keyPath => {
      const val = get(parsedSnapshot.mock, keyPath);
      const target = get(mock, keyPath);
      if (val && target && typeof val === typeof target) {
        set(mock, keyPath, val);
      }
    });
  }

  const snapshot = { className, methodName, mockName, mock };
  const snapshotNameTag = `[${className} ${methodName} ${mockName}]`;

  const result = expect(snapshot).toMatchSnapshot(
    `[mockshot] [${snapshotNameTag}]`
  );
  const pass = result === undefined;
  return { pass };
}

expect.extend({ toMatchMock });

export { toMatchMock };
