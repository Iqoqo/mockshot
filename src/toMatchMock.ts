import { cloneDeep, get, set } from 'lodash';
import pretty from 'json-pretty'
import { generateMocks } from './generateMocks';

declare global {
  namespace jest {
    // tslint:disable-next-line:interface-name
    interface Matchers<R> {
      toMatchMock(className: string, methodName: string, mockName: string, ignoredKeyPaths?: string[]): R
    }
  }
}

expect.addSnapshotSerializer({
  test: val => val.mock,
  print: val => pretty(val)
});

function getSnapshotName(testName, snapshotTag: string) {
  // TODO: If there are more the 1 snapshots (indicated by the number at the end) throw an exception
  return `${testName}: ${snapshotTag} 1`;
}

function getSnapshotTag(className: string, methodName: string, mockName: string) {
  const snapshotNameTag = `[mockshot] [[${className} ${methodName} ${mockName}]]`;

  return snapshotNameTag; 
}

let commonSnapshotState;

afterAll(async() => { 
  if(commonSnapshotState.added > 0 || commonSnapshotState.updated > 0){
    await generateMocks();
  }
});

function toMatchMock(received, className: string, methodName: string, mockName: string, ignoredKeyPaths?: string[], customMatcher?: any) {
  commonSnapshotState = this.snapshotState;

  const snapshotTag = getSnapshotTag(className, methodName, mockName);
  const snapshotName = getSnapshotName(this.currentTestName, snapshotTag);
  const currentSnapshot = this.snapshotState._snapshotData[snapshotName];
  
  if (ignoredKeyPaths && currentSnapshot) {
    const parsedSnapshot = JSON.parse(currentSnapshot);
    received = cloneDeep(received);

    ignoredKeyPaths.forEach(keyPath => {
      const val = get(parsedSnapshot.mock, keyPath);
      const target = get(received, keyPath);
      if(val && target && typeof(val) === typeof(target)){
        set(received, keyPath, val);
      }
    })

  }
  const snapshot = { className, methodName, mockName,  mock: received}
  const snapshotNameTag = `[${className} ${methodName} ${mockName}]`;
  const result = expect(snapshot).toMatchSnapshot(`[mockshot] [${snapshotNameTag}]`);
  const pass = result===undefined;
  return { pass };
}


expect.extend({ toMatchMock });

export { toMatchMock };