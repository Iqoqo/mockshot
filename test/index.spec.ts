import "../src/toMatchMock";
import glob from "glob";
import path from "path";
import Project from "ts-simple-ast";
import fs from "fs";

import { SnapshotState } from "jest-snapshot";
class HelloWorld {
  foo() {
    return;
  }

  bar() {
    return { api: { data: "123" } };
  }
}

describe("toMatchMock", () => {
  beforeAll(() => {});

  it("Should return correct schema", () => {
    const hello = new HelloWorld();
    expect({ foo: "bar" }).toMatchMock(HelloWorld.name, "foo", "success");
  });

  it("Should create another method", () => {
    const hello = new HelloWorld();
    expect(hello.bar()).toMatchMock(HelloWorld.name, "bar", "success");
    expect({error: 'yes'}).toMatchMock(HelloWorld.name, "bar", "no-valid-params");
  });

  it("Should generate mocks", async done => {
    const project = new Project({ compilerOptions: { outDir: "dist/mocks", declaration: true } });

    glob("**/*.ts.mockshot", async function(er, files) {
      // files is an array of filenames.
      // If the `nonull` option is set, and nothing
      // was found, then files is ["**/*.js"]
      // er is an error object or null.
      const mockDef = {} as any;

      files.forEach(file => {
        const absolutePath = path.resolve(file);
        const state = new SnapshotState(absolutePath, {
          snapshotPath: absolutePath
        });
        const data = state._snapshotData;
        const keys = Object.keys(data);

        keys.forEach(async key => {
            if(key.indexOf("[mockshot]")<0){
                return
            }
          const snapshot = JSON.parse(state._snapshotData[key]);
          let classDef = mockDef[snapshot.className];
          if (!classDef) {
            classDef = {};
            mockDef[snapshot.className] = classDef;
          }
          let methodDef = classDef[snapshot.methodName];
          if (!methodDef) {
            methodDef = {};
            classDef[snapshot.methodName] = methodDef;
          }

          let mockName = methodDef[snapshot.mockName];
          if (!mockName) {
            methodDef[snapshot.mockName] = snapshot.mock;
          } else {
            throw Error(
              "Duplicate mock name for method " +
                snapshot.methodName +
                ": " +
                snapshot.mockName
            );
          }
        });
      });

      const classNames = Object.keys(mockDef);
      classNames.forEach(async className => {
        const mockFileName = `mocks/${className}Mock.ts`;

        let myClassFile;
        try {
          await fs.unlinkSync(mockFileName);
        } catch (ex) {
          console.log('No such file', mockFileName);
        }
        myClassFile = project.createSourceFile(mockFileName);

        const classDeclaration = myClassFile.addClass({
          name: className
        });

        const methods = mockDef[className];

        const methodNames = Object.keys(methods);

        methodNames.forEach(methodName => {
            const mocks = methods[methodName];
            const mockNames = Object.keys(mocks);
            let mockTypes = mockNames.map(value => { return `"${value}"`})
            const types = mockTypes.join(' | ')
            console.log('types', types);
          const method = classDeclaration.addMethod({
            isStatic: true,
            parameters: [{name: "mock", type: types}],
            name: methodName,
            returnType: "any"
          });

       

          method.setBodyText(writer =>
            writer.write("switch (mock)").block(() => {
                mockNames.forEach(mockName => {
                    writer.writeLine(`case "${mockName}":`);
                    writer.writeLine(`return ${JSON.stringify(mocks[mockName])}`);
                })
              writer.writeLine(`default:`);
              writer.writeLine(`throw Error("Unknown mock: "+mock);`);
            })
          );
        });
      });

      await project
        .save()

        await project.emit();
        done();
    });
  });
});
