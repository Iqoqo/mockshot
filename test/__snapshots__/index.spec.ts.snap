// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchMock Should create another method: [mockshot] [[HelloWorld bar no-valid-params]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "bar",
  "mock": {
    "error": "yes"
  },
  "mockName": "no-valid-params"
}
`;

exports[`toMatchMock Should create another method: [mockshot] [[HelloWorld bar success]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "bar",
  "mock": {
    "api": {
      "data": "123"
    }
  },
  "mockName": "success"
}
`;

exports[`toMatchMock Should match mock shape: [mockshot] [[HelloWorld shape success]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "shape",
  "mock": {
    "data": {
      "_id": "1234"
    },
    "id": "1234sddd56",
    "some": "value"
  },
  "mockName": "success"
}
`;

exports[`toMatchMock Should return correct schema: [mockshot] [[HelloWorld foo success]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "foo",
  "mock": {
    "foo": "barr"
  },
  "mockName": "success"
}
`;
